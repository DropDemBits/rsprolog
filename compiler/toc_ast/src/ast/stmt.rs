//! Stmt AST Nodes
use crate::ast::expr::{BinaryOp, Expr};
use crate::ast::ident::IdentRef;
use crate::ast::types::Type;
use crate::scope::ScopeBlock;
use toc_core::Location;

/// Kinds of blocks
#[derive(Debug, Copy, Clone)]
pub enum BlockKind {
    /// Main block, root for all blocks, can be an execution block
    Main,
    /// Unit block, cannot be an execution block
    Unit,
    /// Inner block (begin ... end) inside of another block, or general statement block
    InnerBlock,
    /// Function block
    Function,
    /// Procedure block
    Procedure,
    /// Module block
    Module,
    /// Class block
    Class,
    /// Monitor block
    Monitor,
    /// Monitor-Class block, allows both class & monitor statments
    MonitorClass,
}

/// A grouping of statements.
/// Used in multiple statements.
#[derive(Debug, Clone)]
pub struct Block {
    /// Associated scope block
    pub block: ScopeBlock,
    /// Statements as part of the block
    pub stmts: Vec<Stmt>,
}

/// Statement Node Variant
#[derive(Debug, Clone)]
pub enum StmtKind {
    /// No-op statement, does nothing (produced by semi-colons)
    Nop,
    /// Error statement, produced if a valid statement could not be parsed (e.g. in nesting errors or unknown statments)
    Error,
    // Decls
    /// Variable & Constant declaration
    VarDecl {
        /// The identifier(s) declared
        /// If `None`, this statement is a no-op, and only present to allow access to the `type_spec` & `expr`
        idents: Option<Vec<IdentRef>>,
        /// The type spec for all of the identifiers
        type_spec: Option<Box<Type>>,
        /// The (semi-optional) initialization value
        value: Option<Box<Expr>>,
        /// If the declaration is for a const declaration
        is_const: bool,
        /// If the declared identifier(s) should be bound to a register
        bind_to_register: bool,
    },
    /// `type` statement declaration.
    /// The type_spec of `ident` is the declared type
    TypeDecl {
        /// The identifier associated with this type declare.
        /// If `None`, this statement is a no-op, and only allows access to the `resolved_type`
        ident: Option<IdentRef>,
        /// Type associated with the ident
        new_type: Box<Type>,
        /// If the identifier actually declares a new identifier
        is_new_def: bool,
    },
    // Stmts
    /// Simple & Compound assignment expression
    Assign {
        /// The variable reference expression
        var_ref: Box<Expr>,
        /// The (optional) assignment operation
        op: Option<BinaryOp>,
        /// The value to assign
        value: Box<Expr>,
    },
    /// Procedure or function call
    ProcedureCall {
        /// The reference to the procedure or function variable
        proc_ref: Box<Expr>,
    },
    /// Block of statements
    Block { block: Block },
    /// If statement and associated else(if)
    If {
        /// The condition of the if statement
        condition: Box<Expr>,
        /// Link to true branch statement
        true_branch: Box<Stmt>,
        /// Link to false branch statement
        false_branch: Option<Box<Stmt>>,
    },
}

/// Common statement node
#[derive(Debug, Clone)]
pub struct Stmt {
    /// The kind of statement node
    pub kind: StmtKind,
    /// The span of the statement
    pub span: Location,
}

mod pretty_print {
    use super::{Stmt, StmtKind};
    use crate::pretty_print;
    use std::fmt::{self, Write};

    impl fmt::Display for StmtKind {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            match self {
                StmtKind::Nop => f.write_str("<nop>")?,
                StmtKind::Error => f.write_str("<error>")?,
                StmtKind::VarDecl {
                    is_const,
                    bind_to_register,
                    idents,
                    type_spec,
                    value,
                } => {
                    f.write_str(if *is_const { "const " } else { "var " })?;
                    if *bind_to_register {
                        f.write_str("register ")?;
                    }

                    f.write_char('[')?;
                    if let Some(idents) = idents {
                        pretty_print::print_list(f, idents.iter())?;
                    }
                    f.write_char(']')?;

                    if let Some(type_spec) = type_spec {
                        f.write_fmt(format_args!(" : {}", type_spec))?;
                    }

                    if let Some(value) = value {
                        f.write_fmt(format_args!(" := {}", value))?;
                    }
                }
                StmtKind::TypeDecl {
                    ident, new_type, ..
                } => {
                    f.write_str("type [")?;
                    if let Some(ident) = ident {
                        ident.fmt(f)?;
                    }
                    f.write_str("] : ")?;
                    new_type.fmt(f)?;
                }
                StmtKind::Assign { var_ref, op, value } => {
                    if let Some(op) = op {
                        f.write_fmt(format_args!("{0} {1}= {2}", var_ref, op, value))?;
                    } else {
                        f.write_fmt(format_args!("{0} := {1}", var_ref, value))?;
                    }
                }
                StmtKind::ProcedureCall { proc_ref } => proc_ref.fmt(f)?,
                StmtKind::Block { block } => {
                    // use '{' and '}' for block delimiter as it's nicer
                    f.write_char('{')?;

                    if !block.stmts.is_empty() {
                        f.write_char('\n')?;
                    }

                    {
                        let mut indenter = pretty_print::IndentWriter::new(f);

                        for stmt in &block.stmts {
                            indenter.write_fmt(format_args!("{}\n", stmt))?;
                        }
                    }

                    f.write_char('}')?;
                }
                StmtKind::If {
                    condition,
                    true_branch,
                    false_branch,
                } => {
                    f.write_fmt(format_args!("if ({}) then {}", condition, true_branch))?;

                    if let Some(false_branch) = false_branch {
                        f.write_fmt(format_args!("\nelse {}", false_branch))?;
                    }
                }
            }

            Ok(())
        }
    }

    impl fmt::Display for Stmt {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            self.kind.fmt(f)
        }
    }
}
