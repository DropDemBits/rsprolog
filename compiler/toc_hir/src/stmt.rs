//! Statement nodes
use la_arena::Idx;

use crate::{expr, symbol, ty};

pub type StmtIdx = Idx<Stmt>;

#[derive(Debug)]
pub enum Stmt {
    /// Combined representation for `const` and `var` declarations
    /// (disambiguated by `is_const`)
    ConstVar {
        is_register: bool,
        is_const: bool,
        names: Vec<symbol::DefId>,
        type_spec: Option<ty::TypeIdx>,
        init_expr: Option<expr::ExprIdx>,
    },
    // Type { .. },
    // Bind { .. },
    // Proc { .. },
    // Fcn { .. },
    // Process { .. },
    // External { .. },
    // Forward { .. },
    // Deferred { .. },
    // Body { .. },
    // Module { .. },
    // Class { .. },
    // Monitor { .. },
    /// Assignment statement
    /// (also includes compound assignments)
    Assign {
        /// Left hand side of an assignment expression
        lhs: expr::ExprIdx,
        /// Operation performed between the arguments before assignment
        op: AssignOp,
        /// Right hand side of an assignment expression
        rhs: expr::ExprIdx,
    },
    // Open { .. },
    // Close { .. },
    /// Put Statement
    Put {
        /// Stream handle to put the text on.
        /// If absent, should be put on the `stdout` stream.
        stream_num: Option<expr::ExprIdx>,
        /// The items to put out on the stream.
        items: Vec<Skippable<PutItem>>,
        /// If a newline is appended after the all of the put items
        append_newline: bool,
    },
    /// Get Statement
    Get {
        /// Stream handle to get text from.
        /// If absent, should be fetched from the `stdin` stream.
        stream_num: Option<expr::ExprIdx>,
        /// The items to get from the stream.
        items: Vec<Skippable<GetItem>>,
    },
    // Read { .. },
    // Write { .. },
    // Seek { .. },
    // Tell { .. },
    // For { .. },
    // Loop { .. },
    // Exit { .. },
    // If { .. },
    // Case { .. },
    /// Block statement (`begin ... end`)
    Block { stmts: Vec<StmtIdx> },
    // Invariant { .. }
    // Assert { .. }
    // Call { .. }
    // Return { .. }
    // Result { .. }
    // New { .. }
    // Free { .. }
    // Tag { .. }
    // Fork { .. }
    // Signal { .. }
    // Pause { .. }
    // Quit { .. }
    // Break { .. }
    // Checkedness { .. }
    // Pre { .. }
    // Init { .. }
    // Post { .. }
    // Handler { .. }
    // ???: Should the below be real stmts?
    // Inherit { .. }
    // Implement { .. }
    // ImplementBy { .. }
    // Import { .. }
    // Export { .. }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum AssignOp {
    /// Plain assignment
    None,
    /// Addition / Set Union / String Concatenation (`+`)
    Add,
    /// Subtraction / Set Subtraction (`-`)
    Sub,
    /// Multiplication / Set Intersection (`*`)
    Mul,
    /// Integer Division (`div`)
    Div,
    /// Real Division (`/`)
    RealDiv,
    /// Modulo (`mod`)
    Mod,
    /// Remainder (`rem`)
    Rem,
    /// Exponentiation (`**`)
    Exp,
    /// Bitwise/boolean And (`and`)
    And,
    /// Bitwise/boolean Or (`or`)
    Or,
    /// Bitwise/boolean Exclusive-Or (`xor`)
    Xor,
    /// Logical Shift Left (`shl`)
    Shl,
    /// Logical Shift Right (`shr`)
    Shr,
    /// Material Implication (`=>`)
    Imply,
}

impl AssignOp {
    pub fn as_binary_op(self) -> Option<expr::BinaryOp> {
        let op = match self {
            AssignOp::Add => expr::BinaryOp::Add,
            AssignOp::Sub => expr::BinaryOp::Sub,
            AssignOp::Mul => expr::BinaryOp::Mul,
            AssignOp::Div => expr::BinaryOp::Div,
            AssignOp::RealDiv => expr::BinaryOp::RealDiv,
            AssignOp::Mod => expr::BinaryOp::Mod,
            AssignOp::Rem => expr::BinaryOp::Rem,
            AssignOp::Exp => expr::BinaryOp::Exp,
            AssignOp::And => expr::BinaryOp::And,
            AssignOp::Or => expr::BinaryOp::Or,
            AssignOp::Xor => expr::BinaryOp::Xor,
            AssignOp::Shl => expr::BinaryOp::Shl,
            AssignOp::Shr => expr::BinaryOp::Shr,
            AssignOp::Imply => expr::BinaryOp::Imply,
            _ => return None,
        };

        Some(op)
    }
}

/// A generic type representing anything skippable.
/// Only used for text I/O statements (`PutStmt` & `GetStmt`).
#[derive(Debug)]
pub enum Skippable<T> {
    /// This I/O item is skipped.
    Skip,
    /// This I/O item is actually used.
    Item(T),
}

/// A single put item, as part of a put statement.
///
/// If any later optional field is a `Some`, then the earlier fields are
/// guaranteed to also be a `Some` (e.g. if `precision` was `Some`, then
/// `width` is also guaranteed to be `Some`, but not `exponent_width`).
#[derive(Debug)]
pub struct PutItem {
    /// The expression to be put.
    pub expr: expr::ExprIdx,
    /// The minimum printing width.
    pub width: Option<expr::ExprIdx>,
    /// The amount of decimials put for `real*` types.
    pub precision: Option<expr::ExprIdx>,
    /// The minimum printing width for exponents of `real*` types.
    pub exponent_width: Option<expr::ExprIdx>,
    // to maintain doc variant
    _priv: (),
}

impl PutItem {
    /// Constructs a new `PutItem` without any optional arguments.
    pub fn new(expr: expr::ExprIdx) -> Self {
        Self {
            expr,
            width: None,
            precision: None,
            exponent_width: None,
            _priv: (),
        }
    }

    /// Constructs a new `PutItem` with the `width` argument.
    pub fn with_width(expr: expr::ExprIdx, width: expr::ExprIdx) -> Self {
        Self {
            expr,
            width: Some(width),
            precision: None,
            exponent_width: None,
            _priv: (),
        }
    }

    /// Constructs a new `PutItem` with the `precision` argument, and any prerequesite arguments.
    pub fn with_precision(
        expr: expr::ExprIdx,
        width: expr::ExprIdx,
        precision: expr::ExprIdx,
    ) -> Self {
        Self {
            expr,
            width: Some(width),
            precision: Some(precision),
            exponent_width: None,
            _priv: (),
        }
    }

    /// Constructs a new `PutItem` with the `exponent_width` argument, and any prerequesite arguments.
    pub fn with_exponent_width(
        expr: expr::ExprIdx,
        width: expr::ExprIdx,
        precision: expr::ExprIdx,
        exponent_width: expr::ExprIdx,
    ) -> Self {
        Self {
            expr,
            width: Some(width),
            precision: Some(precision),
            exponent_width: Some(exponent_width),
            _priv: (),
        }
    }
}

/// A get item.
#[derive(Debug)]
pub struct GetItem {
    /// The expression to put.
    /// Must be a reference expression.
    pub expr: expr::ExprIdx,
    /// The amount of text to extract.
    pub width: GetWidth,
}

/// The fetch width of a get item.
#[derive(Debug)]
pub enum GetWidth {
    /// Fetch a space delimited portion of text.
    Token,
    /// Fetch a newline terminated portion of text.
    Line,
    /// Fetch a specific amount of characters.
    Chars(expr::ExprIdx),
}
