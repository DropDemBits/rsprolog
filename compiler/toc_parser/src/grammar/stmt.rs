//! Statement grammars
#[cfg(test)]
mod test;

use super::*;

pub(super) fn stmt(p: &mut Parser) -> Option<CompletedMarker> {
    match_token! {
        |p| match {
            TokenKind::Var => { const_var_decl(p) }
            TokenKind::Const => { const_var_decl(p) }
            _ => expr::reference(p).and_then(|m| {
                let m = m.precede(p);
                // check if there's an asn nearby
                if parse_asn_op(p).is_some() {
                    // parse an assign stmt
                    expr::expr(p);

                    Some(m.complete(p,SyntaxKind::AssignStmt))
                } else {
                    // plop as a call stmt
                    Some(m.complete(p,SyntaxKind::CallStmt))
                }
            }),
        }
    }
}

fn parse_asn_op(p: &mut Parser) -> Option<CompletedMarker> {
    let m = p.start();

    if p.eat(TokenKind::Equ) {
        // Simple assignment, but mistyped `=` instead of `:=`
        // TODO: Warn about mistakes like this
        return Some(m.complete(p, SyntaxKind::AsnOp));
    }

    // Reset expected tokens to drop `=`
    p.reset_expected_tokens();

    if p.eat(TokenKind::Assign) {
        // Simple assignment
    } else {
        // Compound assignment
        let valid_asn_op = match_token!(|p| match {
            TokenKind::Imply => { true }
            TokenKind::Or => { true }
            TokenKind::Pipe => { true }
            TokenKind::And => { true }
            TokenKind::Ampersand => { true }
            TokenKind::Plus => { true }
            TokenKind::Minus => { true }
            TokenKind::Xor => { true }
            TokenKind::Star => { true }
            TokenKind::Slash => { true }
            TokenKind::Div => { true }
            TokenKind::Mod => { true }
            TokenKind::Rem => { true }
            TokenKind::Shl => { true }
            TokenKind::Shr => { true }
            TokenKind::Exp => { true }
            _ => false,
        });

        if !valid_asn_op {
            // abandon parsing as compound op
            m.forget(p);
            return None;
        }

        // bump operator
        p.bump();

        if !p.at(TokenKind::Equ) {
            // not a valid asn op, missing equ
            // wrap inside of an error node
            p.error_unexpected_at(m);
            return None;
        }

        // bump equ
        p.bump();
    }

    Some(m.complete(p, SyntaxKind::AsnOp))
}

fn const_var_decl(p: &mut Parser) -> Option<CompletedMarker> {
    assert!(p.at(TokenKind::Var) || p.at(TokenKind::Const));
    let mut require_initializer = p.at(TokenKind::Const);

    let m = p.start();
    p.bump(); // `var` or `const`

    super::name_list(p);

    if p.eat(TokenKind::Colon) {
        // parse type
        p.with_extra_recovery(&[TokenKind::Assign], |p| {
            if ty::ty(p).is_none() {
                // not a type
                p.error();
            }
        });
    } else {
        // initializer is required if type is absent
        require_initializer = true;
    }

    if (require_initializer && p.expect(TokenKind::Assign)) || p.eat(TokenKind::Assign) {
        expr::expr(p);
    }

    Some(m.complete(p, SyntaxKind::ConstVarDecl))
}
