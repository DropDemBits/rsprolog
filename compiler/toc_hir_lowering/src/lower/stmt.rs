//! Lowering into `Stmt` HIR nodes
use toc_hir::{stmt, symbol};
use toc_syntax::ast::{self, AstNode};

impl super::LoweringCtx {
    pub(super) fn lower_stmt(&mut self, stmt: ast::Stmt) -> Option<stmt::StmtIdx> {
        let span = stmt.syntax().text_range();

        let stmt = match stmt {
            ast::Stmt::ConstVarDecl(decl) => self.lower_constvar_decl(decl),
            ast::Stmt::TypeDecl(_) => todo!(),
            ast::Stmt::BindDecl(_) => todo!(),
            ast::Stmt::ProcDecl(_) => todo!(),
            ast::Stmt::FcnDecl(_) => todo!(),
            ast::Stmt::ProcessDecl(_) => todo!(),
            ast::Stmt::ExternalDecl(_) => todo!(),
            ast::Stmt::ForwardDecl(_) => todo!(),
            ast::Stmt::DeferredDecl(_) => todo!(),
            ast::Stmt::BodyDecl(_) => todo!(),
            ast::Stmt::ModuleDecl(_) => todo!(),
            ast::Stmt::ClassDecl(_) => todo!(),
            ast::Stmt::MonitorDecl(_) => todo!(),
            ast::Stmt::AssignStmt(stmt) => self.lower_assign_stmt(stmt),
            ast::Stmt::OpenStmt(_) => todo!(),
            ast::Stmt::CloseStmt(_) => todo!(),
            ast::Stmt::PutStmt(stmt) => self.lower_put_stmt(stmt),
            ast::Stmt::GetStmt(stmt) => self.lower_get_stmt(stmt),
            ast::Stmt::ReadStmt(_) => todo!(),
            ast::Stmt::WriteStmt(_) => todo!(),
            ast::Stmt::SeekStmt(_) => todo!(),
            ast::Stmt::TellStmt(_) => todo!(),
            ast::Stmt::ForStmt(_) => todo!(),
            ast::Stmt::LoopStmt(_) => todo!(),
            ast::Stmt::ExitStmt(_) => todo!(),
            ast::Stmt::IfStmt(_) => todo!(),
            ast::Stmt::CaseStmt(_) => todo!(),
            ast::Stmt::BlockStmt(stmt) => self.lower_block_stmt(stmt),
            ast::Stmt::InvariantStmt(_) => todo!(),
            ast::Stmt::AssertStmt(_) => todo!(),
            ast::Stmt::CallStmt(_) => todo!(),
            ast::Stmt::ReturnStmt(_) => todo!(),
            ast::Stmt::ResultStmt(_) => todo!(),
            ast::Stmt::NewStmt(_) => todo!(),
            ast::Stmt::FreeStmt(_) => todo!(),
            ast::Stmt::TagStmt(_) => todo!(),
            ast::Stmt::ForkStmt(_) => todo!(),
            ast::Stmt::SignalStmt(_) => todo!(),
            ast::Stmt::PauseStmt(_) => todo!(),
            ast::Stmt::QuitStmt(_) => todo!(),
            ast::Stmt::BreakStmt(_) => todo!(),
            ast::Stmt::CheckednessStmt(_) => todo!(),
            ast::Stmt::PreStmt(_) => todo!(),
            ast::Stmt::InitStmt(_) => todo!(),
            ast::Stmt::PostStmt(_) => todo!(),
            ast::Stmt::HandlerStmt(_) => todo!(),
            ast::Stmt::InheritStmt(_) => todo!(),
            ast::Stmt::ImplementStmt(_) => todo!(),
            ast::Stmt::ImplementByStmt(_) => todo!(),
            ast::Stmt::ImportStmt(_) => todo!(),
            ast::Stmt::ExportStmt(_) => todo!(),
            ast::Stmt::PreprocGlob(_) => todo!(),
        }?;

        Some(self.database.stmt_nodes.alloc_spanned(stmt, span))
    }

    fn lower_constvar_decl(&mut self, decl: ast::ConstVarDecl) -> Option<stmt::Stmt> {
        let is_pervasive = decl.pervasive_attr().is_some();
        let is_register = decl.register_attr().is_some();
        let is_const = decl.const_token().is_some();

        let type_spec = decl.type_spec().and_then(|ty| self.lower_type(ty));
        let init_expr = decl.init().map(|expr| self.lower_expr(expr));

        // Declare names after uses to prevent def-use cycles
        let names = self.lower_name_list(decl.decl_list(), is_pervasive)?;

        Some(stmt::Stmt::ConstVar {
            is_register,
            is_const,

            names,
            type_spec,
            init_expr,
        })
    }

    fn lower_assign_stmt(&mut self, stmt: ast::AssignStmt) -> Option<stmt::Stmt> {
        let op = stmt
            .asn_op()
            .and_then(|op| op.asn_kind())
            .map(|op| syntax_to_hir_asn_op(op))
            .unwrap_or(stmt::AssignOp::None);

        let lhs = self.lower_expr(stmt.lhs()?.as_expr());
        let rhs = self.lower_expr(stmt.rhs()?);

        Some(stmt::Stmt::Assign { lhs, op, rhs })
    }

    fn lower_put_stmt(&mut self, stmt: ast::PutStmt) -> Option<stmt::Stmt> {
        let stream_num = self.try_lower_expr(stmt.stream_num().and_then(|s| s.expr()));
        let items = stmt
            .items()
            .filter_map(|item| {
                if item.skip_token().is_some() {
                    Some(stmt::Skippable::Skip)
                } else if let Some(expr) = item.expr() {
                    let expr = self.lower_expr(expr);
                    let width = self.try_lower_expr(item.width().and_then(|o| o.expr()));
                    let precision = self.try_lower_expr(item.fraction().and_then(|o| o.expr()));
                    let exponent_width =
                        self.try_lower_expr(item.exp_width().and_then(|o| o.expr()));

                    let item = match (width, precision, exponent_width) {
                        (Some(width), Some(precision), Some(exponent_width)) => {
                            stmt::PutItem::with_exponent_width(
                                expr,
                                width,
                                precision,
                                exponent_width,
                            )
                        }
                        (Some(width), Some(precision), None) => {
                            stmt::PutItem::with_precision(expr, width, precision)
                        }
                        (Some(width), None, None) => stmt::PutItem::with_width(expr, width),
                        (None, None, None) => stmt::PutItem::new(expr),
                        _ => unreachable!(), // Invariants are being broken
                    };

                    Some(stmt::Skippable::Item(item))
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();
        // Presence means newline should be omitted
        let append_newline = !stmt.range_token().is_some();

        if items.is_empty() {
            // there must be at least one item present
            None
        } else {
            Some(stmt::Stmt::Put {
                stream_num,
                items,
                append_newline,
            })
        }
    }

    fn lower_get_stmt(&mut self, stmt: ast::GetStmt) -> Option<stmt::Stmt> {
        let stream_num = self.try_lower_expr(stmt.stream_num().and_then(|s| s.expr()));
        let items = stmt
            .items()
            .filter_map(|item| {
                if item.skip_token().is_some() {
                    Some(stmt::Skippable::Skip)
                } else if let Some(ref_expr) = item.reference() {
                    let expr = self.lower_expr(ref_expr.as_expr());
                    let width = match item.get_width() {
                        None => stmt::GetWidth::Token,
                        Some(width) if width.star_token().is_some() => stmt::GetWidth::Line,
                        Some(width) => {
                            // Should be an expr for the count
                            stmt::GetWidth::Chars(self.lower_required_expr(width.expr()))
                        }
                    };

                    Some(stmt::Skippable::Item(stmt::GetItem { expr, width }))
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();

        if items.is_empty() {
            // there must be at least one item present
            None
        } else {
            Some(stmt::Stmt::Get { stream_num, items })
        }
    }

    fn lower_block_stmt(&mut self, stmt: ast::BlockStmt) -> Option<stmt::Stmt> {
        self.scopes.push_scope(false);

        let stmts = if let Some(stmts) = stmt.stmt_list() {
            stmts
                .stmts()
                .filter_map(|stmt| self.lower_stmt(stmt))
                .collect()
        } else {
            vec![]
        };

        self.scopes.pop_scope();

        Some(stmt::Stmt::Block { stmts })
    }

    /// Lowers a name list, holding up the invariant that it always contains
    /// at least one identifier
    fn lower_name_list(
        &mut self,
        name_list: Option<ast::NameList>,
        is_pervasive: bool,
    ) -> Option<Vec<symbol::DefId>> {
        let names = name_list?
            .names()
            .filter_map(|name| {
                name.identifier_token().map(|token| {
                    self.scopes.def_sym(
                        token.text(),
                        token.text_range(),
                        symbol::SymbolKind::Declared,
                        is_pervasive,
                    )
                })
            })
            .collect::<Vec<_>>();

        // Invariant: Names list must contain at least one name
        Some(names).filter(|names| !names.is_empty())
    }
}

fn syntax_to_hir_asn_op(op: toc_syntax::AssignOp) -> stmt::AssignOp {
    match op {
        toc_syntax::AssignOp::None => stmt::AssignOp::None,
        toc_syntax::AssignOp::Add => stmt::AssignOp::Add,
        toc_syntax::AssignOp::Sub => stmt::AssignOp::Sub,
        toc_syntax::AssignOp::Mul => stmt::AssignOp::Mul,
        toc_syntax::AssignOp::Div => stmt::AssignOp::Div,
        toc_syntax::AssignOp::RealDiv => stmt::AssignOp::RealDiv,
        toc_syntax::AssignOp::Mod => stmt::AssignOp::Mod,
        toc_syntax::AssignOp::Rem => stmt::AssignOp::Rem,
        toc_syntax::AssignOp::Exp => stmt::AssignOp::Exp,
        toc_syntax::AssignOp::And => stmt::AssignOp::And,
        toc_syntax::AssignOp::Or => stmt::AssignOp::Or,
        toc_syntax::AssignOp::Xor => stmt::AssignOp::Xor,
        toc_syntax::AssignOp::Shl => stmt::AssignOp::Shl,
        toc_syntax::AssignOp::Shr => stmt::AssignOp::Shr,
        toc_syntax::AssignOp::Imply => stmt::AssignOp::Imply,
    }
}
