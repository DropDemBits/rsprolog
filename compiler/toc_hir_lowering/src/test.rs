//! Tests for lowering

use toc_hir::stmt;

#[test]
fn lower_bare_var_def() {
    let parse = toc_parser::parse("var a := b");
    let lowered = crate::lower_ast(parse.syntax());

    let decl = &lowered.database.stmt_nodes.arena[lowered.unit.stmts[0]];
    assert!(matches!(
        decl,
        stmt::Stmt::ConstVar {
            is_pervasive: false,
            is_register: false,
            is_const: false,
            type_spec: None,
            ..
        }
    ))
}

#[test]
fn lower_simple_assignment() {
    let parse = toc_parser::parse("a := b");
    let lowered = crate::lower_ast(parse.syntax());

    let decl = &lowered.database.stmt_nodes.arena[lowered.unit.stmts[0]];
    assert!(matches!(
        decl,
        stmt::Stmt::Assign {
            op: stmt::AssignOp::None,
            ..
        }
    ))
}

#[test]
fn lower_compound_add_assignment() {
    let parse = toc_parser::parse("a += b");
    let lowered = crate::lower_ast(parse.syntax());

    let decl = &lowered.database.stmt_nodes.arena[lowered.unit.stmts[0]];
    assert!(
        matches!(
            decl,
            stmt::Stmt::Assign {
                op: stmt::AssignOp::Add,
                ..
            }
        ),
        "was {:?}",
        decl
    )
}
