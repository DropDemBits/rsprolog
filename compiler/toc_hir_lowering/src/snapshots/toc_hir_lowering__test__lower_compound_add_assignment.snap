---
source: compiler/toc_hir_lowering/src/test.rs
expression: a += b

---
database:
stmt:
Idx::<Stmt>(0) (0..6): Assign(Assign { lhs: Idx::<Expr>(0), op: Spanned(Add, 2..3), rhs: Idx::<Expr>(1) })

expr:
Idx::<Expr>(0) (0..1): Name(Name(UseId(DefId(0), 0)))
Idx::<Expr>(1) (5..6): Name(Name(UseId(DefId(1), 0)))

ty:

root stmts:
[Idx::<Stmt>(0)]
symtab:
DefId(0): (0..1, Symbol { name: "a", kind: Undeclared, is_pervasive: false, def_id: DefId(0), next_use: 1 })
DefId(1): (5..6, Symbol { name: "b", kind: Undeclared, is_pervasive: false, def_id: DefId(1), next_use: 1 })
UseId(DefId(0), 0): 0..1
UseId(DefId(1), 0): 5..6

